File: backend/app/api/schemas/quiz.py
from datetime import datetime
from typing import Any, Literal
import uuid
from pydantic import BaseModel, Field


class QuizQuestionResponse(BaseModel):
    id: uuid.UUID
    position: int
    type: Literal["single", "multiple", "open"]
    text: str
    options: list[str] | None = None
    correct: list[int] | None = None
    rubric: str | None = None
    explanation: str | None = None
    difficulty: str | None = None
    topic: str | None = None
    points: float

    model_config = {"from_attributes": True}


class QuizQuestionStudentResponse(BaseModel):
    id: uuid.UUID
    position: int
    type: Literal["single", "multiple", "open"]
    text: str
    options: list[str] | None = None
    points: float

    model_config = {"from_attributes": True}


class QuizResponse(BaseModel):
    id: uuid.UUID
    lesson_id: uuid.UUID
    version: int
    status: str
    created_at: datetime
    published_at: datetime | None = None
    questions: list[QuizQuestionResponse] = []

    model_config = {"from_attributes": True}


class QuizStudentResponse(BaseModel):
    id: uuid.UUID
    lesson_id: uuid.UUID
    version: int
    status: str
    questions: list[QuizQuestionStudentResponse] = []

    model_config = {"from_attributes": True}


class QuizQuestionUpdate(BaseModel):
    id: uuid.UUID | None = None
    position: int
    type: Literal["single", "multiple", "open"]
    text: str
    options: list[str] = Field(default_factory=list)
    correct: list[int] = Field(default_factory=list)
    rubric: str | None = None
    explanation: str | None = None
    difficulty: str | None = None
    topic: str | None = None
    points: float = 1.0


class QuizUpdateRequest(BaseModel):
    questions: list[QuizQuestionUpdate]


class AnswerSubmission(BaseModel):
    question_id: uuid.UUID
    value: Any


class SaveAnswersRequest(BaseModel):
    answers: list[AnswerSubmission]


class AttemptAnswerResponse(BaseModel):
    question_id: uuid.UUID
    value: Any
    is_correct: bool | None = None
    points: float | None = None
    feedback: str | None = None
    graded_by: str | None = None


class AttemptResponse(BaseModel):
    id: uuid.UUID
    quiz_id: uuid.UUID
    student_id: uuid.UUID
    started_at: datetime
    submitted_at: datetime | None = None
    score: float | None = None
    max_score: float | None = None
    status: str
    answers: list[AttemptAnswerResponse] = []

    model_config = {"from_attributes": True}


class ReportResponse(BaseModel):
    id: uuid.UUID
    lesson_id: uuid.UUID
    quiz_id: uuid.UUID
    stats: dict[str, Any]
    narrative: dict[str, Any] | None = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/quiz.ts
export type QuestionType = 'single' | 'multiple' | 'open';
export type QuizStatus = 'draft' | 'published' | 'archived';

export interface QuizQuestion {
  id: string;
  position: number;
  type: QuestionType;
  text: string;
  options?: string[];
  correct?: number[];
  rubric?: string;
  explanation?: string;
  difficulty?: string;
  topic?: string;
  points: number;
  // stats for live questions
  stats?: {
    answered_count: number;
    total_students: number;
    option_distribution?: number[]; // percentages
    insight?: string;
  };
}

export interface Quiz {
  id: string;
  lesson_id: string;
  version: number;
  status: QuizStatus;
  created_at: string;
  published_at?: string | null;
  questions: QuizQuestion[];
}

export interface AnswerSubmission {
  question_id: string;
  value: any;
}

export interface AttemptAnswer {
  question_id: string;
  value: any;
  is_correct?: boolean;
  points?: number;
  feedback?: string;
  graded_by?: 'auto' | 'vlm' | 'teacher';
}

export interface Attempt {
  id: string;
  quiz_id: string;
  student_id: string;
  started_at: string;
  submitted_at?: string | null;
  score?: number;
  max_score?: number;
  status: 'in_progress' | 'submitted' | 'graded';
  answers: AttemptAnswer[];
}

export interface Report {
  id: string;
  lesson_id: string;
  quiz_id: string;
  average_score: number;
  median_score: number;
  quartiles: [number, number, number]; // [q1, q2, q3]
  at_risk_students: Array<{ student_id: string; score: number; reason: string }>;
  narrative_report: string;
  created_at: string;
}
